### PR TITLE
#6689: add matmul_explicit_same_noc

### PR DESCRIPTION
**Pull request by yugaoTT**
add optimization for matmul_2d, for shapes are DRAM bound.

### CODE REVIEW

File: tt_metal/hw/inc/dataflow_api.h [hunk 1]
```diff
@@ -821,7 +827,7 @@ struct InterleavedAddrGenFast {
     }
 
     FORCE_INLINE
-    void noc_async_read_tile(const uint32_t id, uint32_t dest_addr, const uint32_t offset = 0) const {
+    void noc_async_read_tile(const uint32_t id, uint32_t dest_addr, const uint32_t offset = 0, uint32_t vc = 0) const {
```
Comment (bbradelTT) [inline tt_metal/hw/inc/dataflow_api.h:830]: Why not keep consistent with the other parameters and do `const uint32_t vc = 0`?

Comment (yugaoTT) [reply tt_metal/hw/inc/dataflow_api.h:830]: vc is not constant, we need to change it for different channels

Comment (bbradelTT) [reply tt_metal/hw/inc/dataflow_api.h:830]: ok

**Review by bbradelTT — APPROVED** [review]

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_matmul_1d_2d.py [hunk 2]
```diff
@@ -1194,3 +1194,323 @@ def test_multi_core_matmul_1d_gs(
...
+        # (False, True, True, 256, 256, 4096, None),
+        # (False, True, True, 256, 256, 8192, None),
+        # (False, True, True, 256, 512, 256, None),
+        # (False, True, True, 256, 512, 512, None),
+        # (False, True, True, 256, 512, 1024, None),
+        # (False, True, True, 256, 512, 2048, None),
+        # (False, True, True, 256, 512, 4096, None),
+        # (False, True, True, 256, 512, 8192, None),
+        # (False, True, True, 256, 1024, 256, None),
+        # (False, True, True, 256, 1024, 512, None),
```
Comment (TT-BrianLiu) [inline tests/tt_eager/python_api_testing/unit_testing/misc/test_matmul_1d_2d.py]: Do we need these tests cases? Can we clean it up

File: tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp [hunk 3]
```diff
@@ -64,6 +64,54 @@ uint32_t _get_maximum_block_dim(int32_t block_dim, int32_t in0_block_w) {
...
+        } else if constexpr (std::is_same_v<T, WormholeComputeKernelConfig>) {
+            math_fidelity = compute_kernel_config.math_fidelity;
+        } else {
+            TT_FATAL("arch not supported");
+        }
+    }, compute_kernel_config);
+
+    std::vector<std::vector<int> > bfp8_bfp8_lofi_threshold = { {512, 8192, 1024} };
+    std::vector<std::vector<int> > fp16_fp16_hifi2_threshold = { {256, 8192, 1024}, {512, 256, 4096}, {512, 512, 2048}, {512, 2048, 1024}, {512, 4096, 2048}, {512, 8192, 1024}};
+    std::vector<std::vector<int> > fp16_bfp8_lofi_threshold = { {512, 8192, 2048} };
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp:93]: uint32_t

File: tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp [hunk 4]
```diff
@@ -1172,14 +1221,17 @@ operation::ProgramWithCallbacks Matmul::create_program(
                 );
             }
             else if constexpr (std::is_same_v<ProgramConfigType, MatmulMultiCoreReuseMultiCastProgramConfig>) {
+                bool mcast_use_same_noc = _get_use_same_noc(input_tensor_a, input_tensor_b, this->compute_kernel_config);
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp:1224]: does this need to be here? Can it be inside `matmul_multi_core_reuse_mcast_2d_optimized`? We should try to reduce the number of unnecessary args in our APIs

**Review by TT-BrianLiu — APPROVED** [review]
Please take a look at comments. Functionality looks good but I think we should keep code for matmul as clean as we can. Borys is in the middle of cleaning things up :)

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_matmul_1d_2d.py (hunk 2 above)
Comment (yugaoTT) [reply tests/tt_eager/python_api_testing/unit_testing/misc/test_matmul_1d_2d.py]: These might be required to for later opt, as testing point

Comment (TT-BrianLiu) [reply tests/tt_eager/python_api_testing/unit_testing/misc/test_matmul_1d_2d.py]: ok sounds good

File: tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp (hunk 4 above)
Comment (yugaoTT) [reply tt_eager/tt_dnn/op_library/bmm/bmm_op.cpp:1224]: ok, I will change it

**Pull request merged**